flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/batch/sql/JoinITCase.scala

```diff
@@ -254,7 +254,7 @@ class JoinITCase(
     val table = CollectionDataSets.getSmall3TupleDataSet(env).toTable(tEnv).as('a1, 'a2, 'a3)
     tEnv.registerTable("A", table)
 
-    val sqlQuery2 = "SELECT * FROM (SELECT count(*) FROM A) CROSS JOIN A"
```

Comment:
Yes, when using `COUNT(*) ` we must change result expect. 
``` 
 expected: [3,1,1,Hi, 3,2,2,Hello, 3,3,2,Hello world]
 received: [4,1,1,Hi, 4,2,2,Hello, 4,3,2,Hello world] 
```

Reply:
The current result is correct. `CollectionDataSets.getSmall3TupleDataSet` has three rows. Hence `SELECT count(a1) FROM A` must be `3` and not `4`. 
The PR breaks the correctness of the test.

Reply:
HI, @fhueske I do not think The PR breaks the correctness of the test :), because when we run this test case will match `DataSetAggregateWithNullValuesRule`. which will union a NULL  row to dataset.  so using `COUNT(a1)` result `3` and using `COUNT(*)` result `4`. I think this is batch bug.  The snip as follow:
![image](https://user-images.githubusercontent.com/22488084/34726497-250f4b30-f58f-11e7-92cf-595f9c9a64e0.png) 
I think we can talk about this bug in FlINK-8355... What do you think?

Reply:
The test was working correctly before. If you remove all your changes, the test will pass. It started to fail when you added the count aggregate that is not null-aware. It does not matter which rules are applied during optimization as long as the query computes for the given input data the correct result. The test is correct and is failing due to the changes of the PR. Hence, the PR breaks the correctness of some queries that were working before.

IMO, we cannot merge it in the current state. We have to fix FLINK-8355 before we can merge this PR. I have an idea how to fix FLINK-8355 and will add that to the ticket.

Reply:
Sounds good.